File: src/main/1.2/Largest.c
/*
 * Largest Number Finder
 * Author: [Your Name]
 * Date: [Date]
 * 
 * This program finds the largest number from a list of integers.
 * It supports both predefined arrays and user input.
 */

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

// Function prototypes
int findLargestNumber(int *arr, int count);
void displayArray(int *arr, int count);
void processPresetData();
void processUserInput();
int getUserChoice();

/*
 * findLargestNumber: Finds the maximum value in an integer array
 * Parameters:
 *   arr - pointer to the integer array
 *   count - number of elements in the array
 * Returns: the largest integer found
 */
int findLargestNumber(int *arr, int count) {
    if (count <= 0) {
        printf("Error: Empty array!\n");
        return INT_MIN;
    }
    
    int maxValue = arr[0];  // start with first element
    
    // iterate through remaining elements
    for (int i = 1; i < count; i++) {
        if (arr[i] > maxValue) {
            maxValue = arr[i];
        }
    }
    
    return maxValue;
}

/*
 * displayArray: Prints all elements in the array
 */
void displayArray(int *arr, int count) {
    printf("Numbers in the list: [");
    for (int i = 0; i < count; i++) {
        printf("%d", arr[i]);
        if (i < count - 1) {
            printf(", ");
        }
    }
    printf("]\n");
}

/*
 * processPresetData: Uses a predefined set of numbers
 */
void processPresetData() {
    int testNumbers[] = {45, 23, 89, 12, 67, 34, 91, 56, 78, 21};
    int arraySize = sizeof(testNumbers) / sizeof(testNumbers[0]);
    
    printf("\n--- Preset Data Mode ---\n");
    displayArray(testNumbers, arraySize);
    
    int result = findLargestNumber(testNumbers, arraySize);
    printf("\n%d is the largest number\n\n", result);
}

/*
 * processUserInput: Allows user to enter their own numbers
 */
void processUserInput() {
    int numCount;
    
    printf("\n--- User Input Mode ---\n");
    printf("How many numbers do you want to enter? ");
    
    if (scanf("%d", &numCount) != 1 || numCount <= 0) {
        printf("Invalid input! Please enter a positive number.\n");
        // clear input buffer
        while (getchar() != '\n');
        return;
    }
    
    // dynamically allocate memory for the array
    int *userNumbers = (int *)malloc(numCount * sizeof(int));
    if (userNumbers == NULL) {
        printf("Memory allocation failed!\n");
        return;
    }
    
    // get numbers from user
    printf("Enter %d numbers (separated by spaces or Enter):\n", numCount);
    for (int i = 0; i < numCount; i++) {
        printf("Number %d: ", i + 1);
        if (scanf("%d", &userNumbers[i]) != 1) {
            printf("Invalid input! Please enter an integer.\n");
            free(userNumbers);
            while (getchar() != '\n');
            return;
        }
    }
    
    displayArray(userNumbers, numCount);
    
    int result = findLargestNumber(userNumbers, numCount);
    printf("\n%d is the largest number\n\n", result);
    
    // free allocated memory
    free(userNumbers);
}

/*
 * getUserChoice: Displays menu and gets user selection
 */
int getUserChoice() {
    int choice;
    
    printf("========================================\n");
    printf("   Largest Number Finder Program\n");
    printf("========================================\n");
    printf("1. Use preset data\n");
    printf("2. Enter your own numbers\n");
    printf("3. Exit\n");
    printf("========================================\n");
    printf("Enter your choice (1-3): ");
    
    if (scanf("%d", &choice) != 1) {
        while (getchar() != '\n');  // clear buffer
        return -1;
    }
    
    return choice;
}

/*
 * main: Entry point of the program
 */
int main() {
    int choice;
    int keepRunning = 1;
    
    while (keepRunning) {
        choice = getUserChoice();
        
        switch (choice) {
            case 1:
                processPresetData();
                break;
            case 2:
                processUserInput();
                break;
            case 3:
                printf("\nThank you for using the program. Goodbye!\n");
                keepRunning = 0;
                break;
            default:
                printf("\nInvalid choice! Please select 1, 2, or 3.\n\n");
                break;
        }
    }
    
    return 0;
}

/*
 * Compilation instructions:
 * gcc -o largest_finder largest_finder.c -Wall
 * 
 * Run with:
 * ./largest_finder
 */

File: src/main/1.2/Largest.py
#!/usr/bin/env python3
"""
Largest Number Finder
Author: [Your Name]
Date: [Date]

This program finds the largest number from a list of integers.
It provides multiple ways to input data and display results.
"""

import sys


class NumberAnalyzer:
    """
    A class to handle finding the largest number in a collection.
    This makes the code more organized and reusable.
    """
    
    def __init__(self, numbers=None):
        """Initialize with an optional list of numbers"""
        self.numbers = numbers if numbers else []
    
    def find_largest(self):
        """
        Find and return the largest number in the list.
        Returns None if the list is empty.
        """
        if not self.numbers:
            print("Error: No numbers to analyze!")
            return None
        
        # Start with the first number as our maximum
        max_value = self.numbers[0]
        
        # Check each number to see if it's larger
        for num in self.numbers[1:]:
            if num > max_value:
                max_value = num
        
        return max_value
    
    def display_numbers(self):
        """Display all numbers in a formatted way"""
        if not self.numbers:
            print("No numbers to display.")
            return
        
        print(f"Numbers in the list: {self.numbers}")
        print(f"Total count: {len(self.numbers)}")


def display_menu():
    """Show the main menu options to the user"""
    print("\n" + "=" * 45)
    print("    Largest Number Finder Program")
    print("=" * 45)
    print("1. Use preset sample data")
    print("2. Enter numbers manually")
    print("3. Exit program")
    print("=" * 45)


def get_user_choice():
    """
    Get and validate user's menu choice.
    Returns the choice as an integer, or None if invalid.
    """
    try:
        choice = int(input("Enter your choice (1-3): "))
        if 1 <= choice <= 3:
            return choice
        else:
            print("Please enter a number between 1 and 3.")
            return None
    except ValueError:
        print("Invalid input! Please enter a number.")
        return None


def process_preset_data():
    """Use predefined sample data for demonstration"""
    print("\n--- Preset Sample Data ---")
    sample_numbers = [45, 23, 89, 12, 67, 34, 91, 56, 78, 21]
    
    analyzer = NumberAnalyzer(sample_numbers)
    analyzer.display_numbers()
    
    result = analyzer.find_largest()
    if result is not None:
        print(f"\n{result} is the largest number\n")


def process_manual_input():
    """Let user enter numbers one by one"""
    print("\n--- Manual Input Mode ---")
    
    try:
        count = int(input("How many numbers do you want to enter? "))
        if count <= 0:
            print("Please enter a positive number.")
            return
    except ValueError:
        print("Invalid input! Please enter a valid number.")
        return
    
    numbers = []
    print(f"\nEnter {count} numbers:")
    
    for i in range(count):
        try:
            num = int(input(f"  Number {i + 1}: "))
            numbers.append(num)
        except ValueError:
            print("Invalid input! Skipping this entry.")
            continue
    
    if not numbers:
        print("No valid numbers were entered.")
        return
    
    analyzer = NumberAnalyzer(numbers)
    analyzer.display_numbers()
    
    result = analyzer.find_largest()
    if result is not None:
        print(f"\n{result} is the largest number\n")

def main():
    """
    Main program loop.
    Handles user interaction and program flow.
    """
    print("\nWelcome to the Largest Number Finder!")
    
    # Keep running until user chooses to exit
    while True:
        display_menu()
        choice = get_user_choice()
        
        if choice is None:
            continue
        
        if choice == 1:
            process_preset_data()
        elif choice == 2:
            process_manual_input()
        elif choice == 3:
            print("\nThank you for using the program. Goodbye!\n")
            sys.exit(0)


# This is the entry point of the program
if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        # Handle Ctrl+C gracefully
        print("\n\nProgram interrupted by user. Exiting...")
        sys.exit(0)
    except Exception as e:
        # Catch any unexpected errors
        print(f"\nAn unexpected error occurred: {e}")
        sys.exit(1)


"""
Execution Instructions:
----------------------

Method 1: Direct execution
    python3 largest_finder.py

Method 2: Make it executable (Linux/Mac)
    chmod +x largest_finder.py
    ./largest_finder.py

Features:
---------
- Interactive menu system
- Multiple input methods
- Error handling for invalid inputs
- Clean, organized code structure
- Detailed comments for learning

Notes:
------
- Works on any system with Python 3.x
- No external dependencies required
- Can be easily extended with more features
"""